elements/lisk-p2p/test/unit/peer/base.ts

```diff
@@ -26,45 +51,627 @@ describe('peer', () => {
+						event: p2pPacket.event,
+						data: p2pPacket.data,
+					},
+				);
+			});
+		});
+	});
+
+	describe('#request', () => {
+		it('should throw error if socket does not exists', async () => {
+			const p2pPacket = {
+				data: 'myData',
+				procedure: 'myProcedure',
+			} as P2PRequestPacket;
+			try {
+				await defaultPeer.request(p2pPacket);
+			} catch (e) {
+				expect(e).to.be.an('Error');
+				expect(e.message).to.be.eql('Peer socket does not exist');
+			}
+		});
+
+		it('should emit if socket exists', () => {
+			const p2pPacket = {
+				data: 'myData',
+				procedure: 'myProcedure',
+			} as P2PRequestPacket;
+			const socket = <SCServerSocket>({
+				emit: sandbox.stub(),
+				destroy: sandbox.stub(),
+			} as any);
+			defaultPeer['_socket'] = socket;
+			defaultPeer.request(p2pPacket);
+			expect(socket.emit).to.be.calledOnceWith(REMOTE_SC_EVENT_RPC_REQUEST, {
+				type: '/RPCRequest',
+				procedure: p2pPacket.procedure,
+				data: p2pPacket.data,
+			});
+		});
+	});
+
+	describe('#fetchPeers', () => {
+		it('should call request', async () => {
+			sandbox.stub(defaultPeer, 'request').resolves({
+				data: {
+					peers: [],
+					success: true,
+				},
+			});
+			await defaultPeer.fetchPeers();
+			expect(defaultPeer.request).to.be.calledOnceWith({
+				procedure: REMOTE_EVENT_RPC_GET_PEERS_LIST,
+			});
+		});
+
+		describe('when request() fails', () => {
+			beforeEach(() => {
+				sandbox.stub(defaultPeer, 'emit');
+				sandbox.stub(defaultPeer, 'request').throws('Error');
+			});
+
+			it(`should emit ${EVENT_FAILED_TO_FETCH_PEERS} event`, async () => {
+				try {
+					await defaultPeer.fetchPeers();
+				} catch (e) {
+					expect(defaultPeer.emit).to.be.calledOnceWith(
+						EVENT_FAILED_TO_FETCH_PEERS,
+					);
+				}
+			});
+
+			it('should throw an error', async () => {
+				try {
+					await defaultPeer.fetchPeers();
+				} catch (e) {
+					expect(e).to.be.an.instanceOf(RPCResponseError);
+					expect(e.message).to.be.eql('Failed to fetch peer list of peer');
+					expect(e.peerId).to.be.eql(defaultPeerInfo.ipAddress);
+				}
+			});
+		});
+
+		describe('when request() succeeds', () => {
+			const peers = [
+				{
+					ip: '1.1.1.1',
+					wsPort: 1111,
+					version: '1.1.1',
+				},
+				{
+					ip: '2.2.2.2',
+					wsPort: 2222,
+					version: '2.2.2',
+				},
+			];
+			const sanitizedPeers = [
+				{
+					ipAddress: '1.1.1.1',
+					wsPort: 1111,
+					version: '1.1.1',
+					height: 0,
+					protocolVersion: undefined,
+					os: '',
+				},
+				{
+					ipAddress: '2.2.2.2',
+					wsPort: 2222,
+					version: '2.2.2',
+					height: 0,
+					protocolVersion: undefined,
+					os: '',
+				},
+			];
+
+			beforeEach(() => {
+				sandbox.stub(defaultPeer, 'request').resolves({
+					data: {
+						peers,
+						success: true,
+					},
+				});
+			});
+
+			it('should return a sanitized peer list', async () => {
+				const response = ((await defaultPeer.fetchPeers()) as unknown) as P2PResponsePacket;
```

Comment:
why we are casting it to `P2PResponsePacket`? `defaultPeer.fetchPeers()` returns `ReadonlyArray<P2PPeerInfo>` so we don't have to typecast here